pkg/auth/providers/oidc/oidc_provider.go

```diff
@@ -0,0 +1,353 @@
+		DisplayName:   searchValue,
+		LoginName:     searchValue,
+		PrincipalType: principalType,
+		Provider:      Name,
+	}
+
+	principals = append(principals, p)
+	return principals, nil
+}
+
+func (o *oidcProvider) GetPrincipal(principalID string, token v3.Token) (v3.Principal, error) {
+	var p v3.Principal
+
+	// parsing id to get the external id and type. Exaple oidc_<user|group>://<user sub | group name>
+	var externalID string
+	parts := strings.SplitN(principalID, ":", 2)
+	if len(parts) != 2 {
+		return p, errors.Errorf("invalid id %v", principalID)
+	}
+	externalID = strings.TrimPrefix(parts[1], "//")
+	parts = strings.SplitN(parts[0], "_", 2)
+	if len(parts) != 2 {
+		return p, errors.Errorf("invalid id %v", principalID)
+	}
+
+	principalType := parts[1]
+	if externalID == "" && principalType == "" {
+		return p, fmt.Errorf("[generic oidc]: invalid id %v", principalID)
+	}
+	if principalType != UserType && principalType != GroupType {
+		return p, fmt.Errorf("[generic oidc]: invalid principal type")
+	}
+	if principalID == UserType {
+		p = v3.Principal{
+			ObjectMeta:    metav1.ObjectMeta{Name: principalType + "://" + externalID},
+			DisplayName:   externalID,
+			LoginName:     externalID,
+			PrincipalType: UserType,
+			Provider:      Name,
+		}
+	} else {
+		p = o.groupToPrincipal(externalID)
+	}
+	p = o.toPrincipalFromToken(principalType, p, &token)
+	return p, nil
+}
+
+func (o *oidcProvider) TransformToAuthProvider(authConfig map[string]interface{}) (map[string]interface{}, error) {
+	p := common.TransformToAuthProvider(authConfig)
+	p[publicclient.OIDCProviderFieldRedirectURL] = o.getRedirectURL(authConfig)
+	return p, nil
+}
+
+func (o *oidcProvider) getRedirectURL(config map[string]interface{}) string {
+	return fmt.Sprintf(
+		"%s?client_id=%s&response_type=code&redirect_uri=%s",
+		config["authEndpoint"],
+		config["clientId"],
+		config["rancherUrl"],
+	)
+}
+
+func (o *oidcProvider) RefetchGroupPrincipals(principalID string, secret string) ([]v3.Principal, error) {
+	return nil, errors.New("[generic oidc]: not implemented")
+}
+
+func (o *oidcProvider) CanAccessWithGroupProviders(userPrincipalID string, groupPrincipals []v3.Principal) (bool, error) {
+	config, err := o.getOIDCConfig()
+	if err != nil {
+		logrus.Errorf("[generic oidc]: error fetching OIDCConfig: %v", err)
+		return false, err
+	}
+	allowed, err := o.userMGR.CheckAccess(config.AccessMode, config.AllowedPrincipalIDs, userPrincipalID, groupPrincipals)
+	if err != nil {
+		return false, err
+	}
+	return allowed, nil
+}
+
+func (o *oidcProvider) userToPrincipal(userInfo *oidc.UserInfo, info ClaimInfo) v3.Principal {
+	displayName := info.Name
+	if displayName == "" {
+		displayName = userInfo.Email
+	}
+	p := v3.Principal{
+		ObjectMeta:    metav1.ObjectMeta{Name: Name + "_" + UserType + "://" + userInfo.Subject},
+		DisplayName:   displayName,
+		LoginName:     userInfo.Email,
```

Comment:
with the go-oidc package, you get the generic user info response and on that response it has all of the claims it could retrieve from keycloak. So to get these more specific claim values, I parsed them into the ClaimInfo object so that I could get the user's name - so info.Name is whatever the name value is in keycloak when creating a user. Yes, this does  mean that the user's email will be in the users profile.